auth/design.md

```diff
@@ -0,0 +1,68 @@
+﻿
+# etcd v3 authentication design
+
+## Why not using the existing v2 auth mechanism?
+The v3 protocol uses gRPC as its transport instead of RESTful interface like v2. This new protocol provides an opportunity to iterate on and improve the v2 design. For example, v3 auth has connection based authentication, rather than v2's slower per-request authentication. Additionally, v2 auth's semantics tend to be unwieldy in practice with respect to reasoning about consistency, which will be described in the next sections. For v3, there is a well-defined description and implementation of the authentication mechanism which fixes the deficiencies in the v2 auth system.
+
+### Functionality requirements
+1. per connection authentication, not per request
+   1. User ID + password based authentication implemented on gRPC API
+   1. authentication must be refreshed after auth information update
+1. its functionality should be as simple and useful as v2
+   1. v3 provides a flat key space, unlike the directory structure of v2. Permission check will be provided as interval matching.
+1. it should avoid the consistency problem of v2, described later
+
+### Main required changes
+1. a client needs to create a dedicated connection just for authentication before sending actual requests
+1. add permission information (user ID and authorized revision) to the Raft commands (`etcdserverpb.InternalRaftRequest`)
+1. every request requires permission check in the state machine layer, not in the API layer
+
+### Consistency of the metadata for auth
+The metadata for auth should also be stored and managed in the storage controlled by etcd's Raft protocol like other data stored in etcd. It is required for not sacrificing availability and consistency of the entire etcd cluster. If reading or writing the metadata (e.g. permission information) needs an agreement of every node (more than quorum), single node failure can stop the entire cluster. Requiring all nodes to agree at once means that checking ordinary read/write requests cannot be completed if any cluster member is down, even if the cluster has an available quorum. This unanimous scheme ultimately degrades cluster availability; quorum based consensus from raft should suffice since agreement follows from consistent ordering.
+
+The authentication mechanism in etcd v2 protocol has a tricky part because the metadata consistency described in the above: permission check is processed by the etcd member that receives the client request (etcdserver/etcdhttp/client.go), including follower members. Therefore, it's possible the check may be based on stale metadata.

```

Comment:
s/in etcd v2 protocol/in the etcd v2 protocol/
s/metadata consistency described in the above:/metadata consistency should work as in the above, but does not:/
s/permission check is processed/each permission check is processed
etcdserver/etcdhttp/client.go -> etcdserver/api/v2http/client.go